chore: introduce API changes to ease support of zerolog.

jcchavezs opened the pull request:
This PR aims to ease the introduction of zerolog in one hand to still use coraza types but let the API interact with zerolog api aswell.

Ping @anuraaga

manojgop committed:
Use zerolog for logging.

zerolog is optimized for logging and does not allocate memory unlike fmr.Sprintf()
for debug logs

Signed-off-by: Manoj Gopalakrishnan <manoj.gopalakrishnan@intel.com>

anuraaga committed:
Remove MatchData.IsNil (#677)

jcchavezs committed:
tests: adds more coverage for CTL action. (#679)

* tests: adds more coverage for CTL action.

* tests: improves naming.

anuraaga committed:
Avoid intermediate string when formatting variables for debug (#675)

jcchavezs committed:
chore: introduce API changes to ease support of zerolog.

ns-sundar commented on actions/ctl.go:
```diff
@@ -103,7 +106,10 @@ func (a *ctlFn) Evaluate(_ rules.RuleMetadata, txS rules.TransactionState) {
-			tx.WAF.Logger.Error("[ctl:AuditEngine] %s", err.Error())
+			tx.WAF.Logger.Error().
+				Str("ctl", "AuditEngine").
+				Str("value", a.value).
```
- The `a.value` was not in the original log statement, but seems like a worthwhile add here.
- Perhaps we could consider a uniform format where every log ends with a Msg() for uniformity? Just a thought.

jcchavezs replied:
Yeah, makes sense. I just did not know what to use as msg but let me try something.

ns-sundar commented on actions/ctl.go:
```diff
@@ -113,36 +119,52 @@ func (a *ctlFn) Evaluate(_ rules.RuleMetadata, txS rules.TransactionState) {
 		if err != nil {
-			tx.WAF.Logger.Error("[ctl:RuleEngine] %s", err.Error())
+			tx.WAF.Logger.Error().
+				Str("ctl", "RuleEngine").
```
Why not log `a.value` here too?

jcchavezs replied:
Forgot to.

ns-sundar commented on actions/ctl.go:
```diff
@@ -113,36 +119,52 @@ func (a *ctlFn) Evaluate(_ rules.RuleMetadata, txS rules.TransactionState) {
 		if err != nil {
-			tx.WAF.Logger.Error("[ctl:RuleRemoveByID] %s", err.Error())
+			tx.WAF.Logger.Error().
+				Str("ctl", "RuleRemoveByID").
```
Why not log `a.value` here too?

jcchavezs replied:
Added.

anuraaga commented on actions/setvar.go:
```diff
@@ -92,15 +96,23 @@ func (a *setvarFn) evaluateTxCollection(r rules.RuleMetadata, tx rules.Transacti
 			sum, err = strconv.Atoi(value[1:])
 			if err != nil {
-				tx.DebugLogger().Error("[%s] Invalid value for setvar %q on rule %d", tx.ID(), value, r.ID())
+				tx.DebugLogger().Error().
```
I think all the logs should have a Msg, eg invalid value for this one

jcchavezs replied:
Done

ns-sundar commented on actions/expirevar.go:
```diff
@@ -47,7 +47,7 @@ func (a *expirevarFn) Init(_ rules.RuleMetadata, data string) error {
 	// Not supported
 	// TODO(jcchavezs): Shall we log a message?
-	// tx.WAF.Logger.Error("Expirevar was used but it's not supported", zap.Int("rule", r.Id))
+	// tx.WAF.Logger.Error().Msg("Expirevar was used but it's not supported", zap.Int("rule", r.Id))
```
What is the reason to retain a commented log?

jcchavezs replied:
Good question, let's remove it in another PR.

anuraaga reviewed (commented):
I think it will be good if the default still uses the standard library and we can provide a plugin to implement with zerolog, or zap or whatever. Does this PR achieve that?

manojgop commented on internal/corazawaf/logger.go:
```diff
@@ -2,99 +2,3 @@
 // SPDX-License-Identifier: Apache-2.0
 
 package corazawaf
```
Looks like this file can be deleted ?

jcchavezs replied:
Done.

manojgop commented:
@jcchavezs Thanks for raising this PR.  Some checks are failing due to incorrect conversion between integer types

manojgop commented on internal/seclang/rule_parser.go:
```diff
@@ -447,9 +447,9 @@ actionLoop:
 				continue actionLoop
 			}
-			cval.WriteString(string(c))
+			cval.WriteRune(c)
```
Looks good. But this change is unrelated to zerolog and was initially added in this commit -> https://github.com/manojgop/coraza/pull/2/commits/187daf6cf954f3d79feaf35c18e8e9a12e549e59

manojgop commented:
> I think it will be good if the default still uses the standard library and we can provide a plugin to implement with zerolog, or zap or whatever. Does this PR achieve that?

This is a good suggestion. However, zerolog uses a different API for log message formatting compared to log package in standard library. We need to avoid invoking fmt.Sprintf(message, args..) variadic function if Debug is disabled to avoid heap allocations of arguments by go compiler escape analysis. zerolog APIs are designed to avoid heap allocations and reflection. It could be tricky to use a generic API and to format the log message that could satisfy both zerolog and standard log library APIs and at the same time avoid heap allocations if Debug logs is disabled.

anuraaga commented:
> It could be tricky to use a generic API and to format the log message that could satisfy both zerolog and standard log library APIs and at the same time avoid heap allocations if Debug logs is disabled.

The zerolog version can be the optimized version, while it is good to work out of the box with the standard log library that can have extra heap allocations as a tradeoff.

Notably, being able to implement with zap does seem important - for example Caddy uses zap for logging, and while it's not the end of the world to use a different library with Coraza I think it'd generally be better to be able to plug zap logger in for that connector. So I'd like to make sure the logging implementation is decoupled from the logger plugin interface if at all possible.

manojgop commented:
Just FYI - A related change in coraza-proxy-wasm to use zerolog is https://github.com/manojgop/coraza-proxy-wasm/commit/e2cd02f9b9f80d5380271ecbf2c8038e003e7c9f. This would need modifications as well

jcchavezs commented:
I think one idea would be to mimic the zerolog approach and have a similar
API which can work with standard logger e.g. doing `message key1=value1
key2=value2` for std logger which is plain.

While sticking with zerolog does not seem too crazy, I think @anuraaga made
a point with caddy (using zap) and actually proxy-wasm has an own logger
API which can fit in this API aswell.

On Wed, 1 Mar 2023, 06:39 manojgop, ***@***.***> wrote:

> Just FYI - A related change in coraza-proxy-wasm to use zerolog is
> ***@***.***
> <https://github.com/manojgop/coraza-proxy-wasm/commit/e2cd02f9b9f80d5380271ecbf2c8038e003e7c9f>.
> This would need modifications as well
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/manojgop/coraza/pull/3#issuecomment-1449381017>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYAVGGGEKN2NX3H3Y7FLWZ3OIZANCNFSM6AAAAAAVLJZQWA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

manojgop committed:
Use string builder in parseActions() to optimize memory usage. (#681)

Use string builder in parseActions() to optmize memory usage.

Use string builder in parseActions() function to efficiently concatenate
strings and optimize memory usage. go bench profile (testing/coreruleset) shows
less allocations in seclang.parseActions function.

Signed-off-by: Manoj Gopalakrishnan <manoj.gopalakrishnan@intel.com>

jcchavezs committed:
chore: uses interface of logging.

jcchavezs commented on loggers/debug.go:
```diff
@@ -4,34 +4,56 @@
+func (e NopEvent) Stringer(key string, val fmt.Stringer) Event { return e }
+
 // DebugLogger is used to log SecDebugLog messages
 type DebugLogger interface {
```
How about this logger interface? @anuraaga @manojgop @ns-sundar

jcchavezs commented on loggers/default.go:
```diff
@@ -0,0 +1,155 @@
+	return e
+}
+
+type defaultLogger struct {
```
This is the default implementation based on standard `log` package cc @manojgop @anuraaga @ns-sundar

jcchavezs committed:
chore: uniformize logger messages.

jcchavezs commented:
I gave some turnarounds to the API. Please do review and let me know if this feels right so I can backfill tests.

manojgop commented:
> I gave some turnarounds to the API. Please do review and let me know if this feels right so I can backfill tests.

Could you please check that heap allocation is not there in Debug Log statements (if Debug logs are disabled) if we use zerolog instead of default log. Could you elaborate how can we replace default log package with zerolog (at compile time as a plugin) ?

manojgop commented:
> > I gave some turnarounds to the API. Please do review and let me know if this feels right so I can backfill tests.
> 
> Could you please check that heap allocation is not there in Debug Log statements (if Debug logs are disabled) if we use zerolog instead of default log. Could you elaborate how can we replace default log package with zerolog (at compile time as a plugin) ?

Got it. I see that APIs are designed similar to zerolog APIs and with that Debug() log statements will avoid heap allocations if Debug log is disabled. Overall it looks good to me :-)

jcchavezs committed:
feat: polish debug logger API.

jcchavezs committed:
chore: small nit in nop event.

jcchavezs committed:
chore: uniformize log fields.

jcchavezs commented:
I think the API is ready @anuraaga @ns-sundar @manojgop. If we merge this now then @manojgop can open a PR against coraza repo itself I guess.

![image](https://user-images.githubusercontent.com/3075074/222267109-b9bc69da-8925-4a83-b5e7-18a0a3d49f82.png)

doEvaluate() function heap allocation has reduced

![image](https://user-images.githubusercontent.com/3075074/222267133-5708ba0b-8b9e-4e0e-8c89-b7ea8c84cdf6.png)

Log statements are not allocating memory

![image](https://user-images.githubusercontent.com/3075074/222267185-01460eef-5dd4-4f55-be2c-bc902ed75e25.png)

Thanks @manojgop for the screenshots.

jcchavezs commented on debuglogger/logger.go:
```diff
@@ -0,0 +1,61 @@
+// Logger is used to log SecDebugLog messages
+// This interface is highly inspired in github.com/rs/zerolog logger and the aim
+// is to avoid allocations while logging.
+type Logger interface {
```
Logger implementation would follow this interface, being zap or proxywasm cc @anuraaga

jcchavezs committed:
docs: explains the timestamp skip.

anuraaga commented on actions/ctl.go:
```diff
@@ -113,36 +120,54 @@ func (a *ctlFn) Evaluate(_ rules.RuleMetadata, txS rules.TransactionState) {
 		limit, err := strconv.ParseInt(a.value, 10, 64)
 		if err != nil {
-			tx.WAF.Logger.Error("[ctl:RequestBodyLimit] Incorrect integer CTL value %q", a.value)
+			tx.WAF.Logger.Error().
```
This is missing `Msg` so won't be outputed - it seems that zerolog has `Send()` for msg-less logs.

Note that this is the trickiness of the zerolog API that achieves minimum allocations with the chance of not actually logging by mistake - it's definitely much harder to use than something like zap. But I guess we can deal with it. If there's not a linter for it yet I guess we could write it

anuraaga replied:
Ah nice, there's an official one

https://github.com/rs/zerolog/blob/master/cmd/lint/README.md

jcchavezs replied:
Lint sounds good. BTW, one of the things I have been doing lately is to cover logs with unit tests but linter seems easy. Let me open an issue for that.

anuraaga commented on debuglogger/default.go:
```diff
@@ -0,0 +1,159 @@
+		return
+	}
+
+	s := strings.Builder{}
```
Think `e.logger.Printf` would be better than using this builder to avoid the intermediate string.

`e.logger.Printf("[%s] %s%s\n", e.level.String(), msg, e.fields)`

anuraaga commented on debuglogger/default.go:
```diff
@@ -0,0 +1,159 @@
+}
+
+func (l defaultLogger) Trace() Event {
+	if l.level < LogLevelTrace || l.Logger == nil {
```
Don't think `Logger` can be `nil`, especially after making it a private field as it probably should be

jcchavezs replied:
It was in NOP case but I just changed it based on this convenience https://cs.opensource.google/go/go/+/refs/tags/go1.20.1:src/log/log.go;l=72

anuraaga reviewed (commented):
Seems reasonable to me

ns-sundar commented:
> > It could be tricky to use a generic API and to format the log message that could satisfy both zerolog and standard log library APIs and at the same time avoid heap allocations if Debug logs is disabled.

Do we necessarily need two logging frameworks? There may be ways to switch to Zerolog that may work with the current log statements, so that we don't need a forklift upgrade.

> The zerolog version can be the optimized version, while it is good to work out of the box with the standard log library that can have extra heap allocations as a tradeoff.

The standard log library does not provide structured logging, unless you have something else in mind. So, having both will result in inconsistent logging.

As a data point, [Crowdstrike switched to zerolog](https://www.crowdstrike.com/blog/logging-with-go/) in full and saw good benefits. 

> Notably, being able to implement with zap does seem important - for example Caddy uses zap for logging, and while it's not the end of the world to use a different library with Coraza I think it'd generally be better to be able to plug zap logger in for that connector. So I'd like to make sure the logging implementation is decoupled from the logger plugin interface if at all possible.

This is a valid point. But, first, zerolog is [widely used](https://github.com/rs/zerolog/wiki/Who-uses-zerolog). Secondly, [zap itself says that zerolog is faster](https://github.com/uber-go/zap#performance). That said, if you are advocating some kind of neutral interface that will work with multiple loggers, for future-proofing, I am ok -- just wondering what effort that will entail.

anuraaga commented:
> just wondering what effort that will entail.

I think @jcchavezs has done most of the effort for it here, we already had a neutral interface for loggers and just have tweaked it now to enable a zero allocation backend.

As a core library rather than an app, we try to make it generic so that it is those integrating with it that can make their decisions. If apps can pick zerolog, zap, or even std library based on their existing dependencies or lack of dependencies, it makes Coraza a more useful tool in the toolbox than if we force a decision on it.

jcchavezs committed:
chore: improves nop loggic and also print allocations.

jcchavezs committed:
chore: fixes missing log messages.

jcchavezs committed:
chore: generic improvements.

jcchavezs committed:
tests: improves coverage.

jcchavezs committed:
chore: readds log.

jcchavezs committed:
tests: improves coverage.

jcchavezs committed:
feat: adds logger per transaction to allow 'ctlDebugLogLevel'.

jcchavezs committed:
chore: nit printf.

jcchavezs committed:
tests: covers 'ctlDebugLogLevel'.

anuraaga committed:
Enable xml processor on TinyGo (#685)

jcchavezs committed:
feat: adds Logger.With context fields as per @ns-sundar suggestion.

jcchavezs committed:
chore: removes Event.Errs method to narrow down the API surface.

jcchavezs committed:
Merge branch 'v3/dev' into use_zerolog_frictionless

jcchavezs committed:
docs: adds more comments on the IsEnabled method.

jcchavezs committed:
chore: uses contextful logs for rule.

jcchavezs committed:
feat: adds PrinterFactory abstraction to ease the plug of new logs.

jcchavezs committed:
chore: refine printer.

jcchavezs committed:
chore: rollsback creating a new log per rule evaluation.
